[SPARK-1855] Local checkpointing

andrewor14 opened the pull request:
Certain use cases of Spark involve RDDs with long lineages that must be truncated periodically (e.g. GraphX). The existing way of doing it is through `rdd.checkpoint()`, which is expensive because it writes to HDFS. This patch provides an alternative to truncate lineages cheaply _without providing the same level of fault tolerance_.

**Local checkpointing** writes checkpointed data to the local file system through the block manager. It is much faster than replicating to a reliable storage and provides the same semantics as long as executors do not fail. It is accessible through a new operator `rdd.localCheckpoint()` and leaves the old one unchanged. Users may even decide to combine the two and call the reliable one less frequently.

The bulk of this patch involves refactoring the checkpointing interface to accept custom implementations of checkpointing. [Design doc](https://issues.apache.org/jira/secure/attachment/12741708/SPARK-7292-design.pdf).

Commit:
Refactor checkpoint interface for modularity

This commit makes two classes abstract: `RDDCheckpointData` and
`CheckpointRDD`. It implements the existing fault-tolerant
checkpointing by subclassing these abstract classes. The goal
of this commit is to retain as much functionality as possible.
Much of the code is just moved from one file to another.

The following commits will add an implementation for unreliable
checkpointing.

Commit:
Fix tests

The parent base class was not serializable while the child is,
causing some java invocation exception. Also, a prior code clean
up caused an array out of bounds exception, which is now fixed
in this commit.

Commit:
First implementation of local checkpointing

The write path runs a job to put each partition into disk store,
while the read path simply reads these blocks back from the disk
store.

Commit:
Rename a few methods with awkward names (minor)

Commit:
Clean local checkpoint files through RDD cleanups

This commit simplifies the previous one in removing the special
`LocalCheckpointBlockId`, which is not needed if we use the
checkpoint RDD's ID instead of the parent RDD's. This allows us
to simply reuse the RDD cleanup code path, which is nice.

Commit:
Refactor CheckpointSuite to test local checkpointing

This commit makes each test in CheckpointSuite run twice, once
for normal checkpointing and another time for local checkpointing.
This commit also fixes legitimate test failures after the
refactoring.

Commit:
Add local checkpoint clean up tests

Commit:
Add a few warnings and clear exception messages

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36749 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36749/consoleFull) for   PR 7279 at commit [`1324d25`](https://github.com/apache/spark/commit/1324d25f1b0e6279ac5c317e0ecb69d26d76ed3c).

SparkQA commented:
[Test build #36749 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36749/console) for   PR 7279 at commit [`1324d25`](https://github.com/apache/spark/commit/1324d25f1b0e6279ac5c317e0ecb69d26d76ed3c).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

witgo commented:
This is very cool PR.

Commit:
Fix style

andrewor14 commented:
Thank you @witgo. :)

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36756 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36756/consoleFull) for   PR 7279 at commit [`6602052`](https://github.com/apache/spark/commit/66020529cc22d0aa677a547652b53653ce618f38).

SparkQA commented:
[Test build #36756 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36756/console) for   PR 7279 at commit [`6602052`](https://github.com/apache/spark/commit/66020529cc22d0aa677a547652b53653ce618f38).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Fix mima?

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36761 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36761/consoleFull) for   PR 7279 at commit [`d980757`](https://github.com/apache/spark/commit/d98075785c60e2baa50abebda8c500da4d2c09f0).

SparkQA commented:
[Test build #36761 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36761/console) for   PR 7279 at commit [`d980757`](https://github.com/apache/spark/commit/d98075785c60e2baa50abebda8c500da4d2c09f0).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Fix mima

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36815 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36815/consoleFull) for   PR 7279 at commit [`ee8e85e`](https://github.com/apache/spark/commit/ee8e85ec24c4d546265b8b184b037cc03805c507).

SparkQA commented:
[Test build #36815 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36815/console) for   PR 7279 at commit [`ee8e85e`](https://github.com/apache/spark/commit/ee8e85ec24c4d546265b8b184b037cc03805c507).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Fix style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36820 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36820/consoleFull) for   PR 7279 at commit [`125af6f`](https://github.com/apache/spark/commit/125af6f02f840e41e6d3f31cbef7e59eec13efd1).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36826 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36826/consoleFull) for   PR 7279 at commit [`03f3126`](https://github.com/apache/spark/commit/03f3126da2d0c3aa21acc1f88d5ae9f631c0ae99).

SparkQA commented:
[Test build #36826 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36826/console) for   PR 7279 at commit [`03f3126`](https://github.com/apache/spark/commit/03f3126da2d0c3aa21acc1f88d5ae9f631c0ae99).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36836 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36836/consoleFull) for   PR 7279 at commit [`3d4a717`](https://github.com/apache/spark/commit/3d4a7176e7a3aa6f18035138689075516327e7e5).

SparkQA commented:
[Test build #36836 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36836/console) for   PR 7279 at commit [`3d4a717`](https://github.com/apache/spark/commit/3d4a7176e7a3aa6f18035138689075516327e7e5).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #36820 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36820/console) for   PR 7279 at commit [`125af6f`](https://github.com/apache/spark/commit/125af6f02f840e41e6d3f31cbef7e59eec13efd1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Simplify LocalCheckpointRDD + docs + clean ups

This commit does several things:

(1) First, LocalCheckpointRDD is made significantly simplier.
Instead of fetching block IDs from everyone and verifying whether
the partition indices are continuous, we simply use the original
RDD's partition indices.

(2) Many checkpoint-related methods are now documented, and failure
conditions in local checkpointing now present more informative
error messages.

(3) General code clean ups (reordering things for readability)

Commit:
Rename a few more awkwardly named methods (minor)

Commit:
Add fine-grained tests for local checkpointing

This augments the existing end-to-end tests in CheckpointSuite.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36882 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36882/consoleFull) for   PR 7279 at commit [`5da18c7`](https://github.com/apache/spark/commit/5da18c7c2de5745f19404c29fdca170e0dc29a61).

SparkQA commented:
[Test build #36882 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36882/console) for   PR 7279 at commit [`5da18c7`](https://github.com/apache/spark/commit/5da18c7c2de5745f19404c29fdca170e0dc29a61).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

Commit:
Fix style

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36885 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36885/consoleFull) for   PR 7279 at commit [`c449b38`](https://github.com/apache/spark/commit/c449b38f420e07e581541933c53060f319f948ec).

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #1017 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1017/consoleFull) for   PR 7279 at commit [`c449b38`](https://github.com/apache/spark/commit/c449b38f420e07e581541933c53060f319f948ec).

SparkQA commented:
[Test build #1018 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1018/consoleFull) for   PR 7279 at commit [`c449b38`](https://github.com/apache/spark/commit/c449b38f420e07e581541933c53060f319f948ec).

SparkQA commented:
[Test build #1017 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1017/console) for   PR 7279 at commit [`c449b38`](https://github.com/apache/spark/commit/c449b38f420e07e581541933c53060f319f948ec).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #36885 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36885/console) for   PR 7279 at commit [`c449b38`](https://github.com/apache/spark/commit/c449b38f420e07e581541933c53060f319f948ec).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #1018 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1018/console) for   PR 7279 at commit [`c449b38`](https://github.com/apache/spark/commit/c449b38f420e07e581541933c53060f319f948ec).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

Commit:
Merge branch 'master' of github.com:apache/spark into local-checkpoint

Conflicts:
	project/MimaExcludes.scala

Commit:
Express local checkpointing through caching the original RDD

This allows us to run local checkpointing using MEMORY_AND_DISK
instead of just DISK_ONLY. The motivation is primarily for
performance.

One potential complexity is that it adds more things into
memory store, which may change the LRU behavior for certain
workloads. This happens, for instance, if the RDD action did
not fully drain the iterator (e.g. take).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37169 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37169/consoleFull) for   PR 7279 at commit [`db70dc2`](https://github.com/apache/spark/commit/db70dc23262bc7baab8d4eacbbf6df0ef2187306).

SparkQA commented:
**[Test build #37169 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37169/console)**     for PR 7279 at commit [`db70dc2`](https://github.com/apache/spark/commit/db70dc23262bc7baab8d4eacbbf6df0ef2187306)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37204 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37204/consoleFull) for   PR 7279 at commit [`db70dc2`](https://github.com/apache/spark/commit/db70dc23262bc7baab8d4eacbbf6df0ef2187306).

SparkQA commented:
**[Test build #37204 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37204/console)**     for PR 7279 at commit [`db70dc2`](https://github.com/apache/spark/commit/db70dc23262bc7baab8d4eacbbf6df0ef2187306)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into local-checkpoint

Commit:
Avoid traversing dependency tree + rewrite tests

This commit fixes previous test failures in GraphX and Bagel,
where tests involving long dependency chains would time out.
This is because the previous implementation unconditionally
traversed the entire dependency tree every time before a job
is run, which is not safe if there are circular dependencies.

This commit changes the storage level mutation behavior such
that calling localCheckpoint() with persist(...) in any order
will yield the right storage level. LocalCheckpointSuite is
completely rewritten in a way that ensures this does not
break crucial functionality.

Commit:
Simplify LocalCheckpointRDD

It doesn't even need to fetch blocks because the child should
already do this through getOrCompute. This commit makes it simply
a placeholder for providing error messages when things fail.

Commit:
Merge branch 'master' of github.com:apache/spark into local-checkpoint

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37306 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37306/consoleFull) for   PR 7279 at commit [`4eb6eb1`](https://github.com/apache/spark/commit/4eb6eb1af5595ed68025fd68ee3ed4dbea881591).

SparkQA commented:
[Test build #37306 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37306/console) for   PR 7279 at commit [`4eb6eb1`](https://github.com/apache/spark/commit/4eb6eb1af5595ed68025fd68ee3ed4dbea881591).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Exchange(newPartitioning: Partitioning, child: SparkPlan) extends UnaryNode `

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
@tdas can you start looking at this?

andrewor14 commented on core/src/main/scala/org/apache/spark/rdd/ReliableCheckpointRDD.scala:
```diff
@@ -0,0 +1,194 @@
+  /**
+   * Write this partition's values to a checkpoint file.
+   */
+  def writeCheckpointFile[T: ClassTag](
```
These are all moved from `CheckpointRDD`. I just renamed some of the methods.

Commit:
Merge branch 'master' of github.com:apache/spark into local-checkpoint

Commit:
Update a few comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #1079 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1079/consoleFull) for   PR 7279 at commit [`a92657d`](https://github.com/apache/spark/commit/a92657d815e7837a64d69546acc954a792ae1d1a).

SparkQA commented:
[Test build #1080 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1080/consoleFull) for   PR 7279 at commit [`a92657d`](https://github.com/apache/spark/commit/a92657d815e7837a64d69546acc954a792ae1d1a).

SparkQA commented:
[Test build #1081 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1081/consoleFull) for   PR 7279 at commit [`a92657d`](https://github.com/apache/spark/commit/a92657d815e7837a64d69546acc954a792ae1d1a).

SparkQA commented:
[Test build #1080 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1080/console) for   PR 7279 at commit [`a92657d`](https://github.com/apache/spark/commit/a92657d815e7837a64d69546acc954a792ae1d1a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1081 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1081/console) for   PR 7279 at commit [`a92657d`](https://github.com/apache/spark/commit/a92657d815e7837a64d69546acc954a792ae1d1a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1079 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1079/console) for   PR 7279 at commit [`a92657d`](https://github.com/apache/spark/commit/a92657d815e7837a64d69546acc954a792ae1d1a).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1089 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1089/consoleFull) for   PR 7279 at commit [`a92657d`](https://github.com/apache/spark/commit/a92657d815e7837a64d69546acc954a792ae1d1a).

SparkQA commented:
[Test build #1091 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1091/consoleFull) for   PR 7279 at commit [`a92657d`](https://github.com/apache/spark/commit/a92657d815e7837a64d69546acc954a792ae1d1a).

SparkQA commented:
[Test build #1091 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1091/console) for   PR 7279 at commit [`a92657d`](https://github.com/apache/spark/commit/a92657d815e7837a64d69546acc954a792ae1d1a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1089 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1089/console) for   PR 7279 at commit [`a92657d`](https://github.com/apache/spark/commit/a92657d815e7837a64d69546acc954a792ae1d1a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

Commit:
Use MEMORY_AND_DISK as the default local checkpoint level

For performance!

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37549 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37549/consoleFull) for   PR 7279 at commit [`f5be0f3`](https://github.com/apache/spark/commit/f5be0f37c6e9941ad03df053072f9c2b0d4fbffa).

SparkQA commented:
[Test build #37549 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37549/console) for   PR 7279 at commit [`f5be0f3`](https://github.com/apache/spark/commit/f5be0f37c6e9941ad03df053072f9c2b0d4fbffa).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Fix tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37568 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37568/consoleFull) for   PR 7279 at commit [`e908a42`](https://github.com/apache/spark/commit/e908a42c698d295e78ffd65f42b1c3ca97e9f46d).

SparkQA commented:
[Test build #37568 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37568/console) for   PR 7279 at commit [`e908a42`](https://github.com/apache/spark/commit/e908a42c698d295e78ffd65f42b1c3ca97e9f46d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
Hi all, I spent some time benchmarking the performance improvements of this feature. I ran the following on a 4-node `m1.xlarge` cluster with 6 executors, each with `4g` of memory. The numbers reported here are the median over 10 runs.

**Micro-benchmark.** Checkpoint a ~700MB input RDD
- normal checkpoint: 9337ms
- local checkpoint: 1599ms
- improvement: *82.9%*

**Macro-benchmark.** Run EM LDA on a ~235MB file on HDFS
- `--maxIterations` = 20
- `--checkpointInterval` = 2
- `--k` = 10
- normal checkpoint: 96078ms
- local checkpoint: 79592ms
- improvement: *17.2%*

The performance improvement in LDA is not as dramatic as the micro benchmark. This is because the majority of time is spent on actual computation rather than checkpointing.

witgo commented:
This is a very good performance for ML.    I will test this feature on [cloudml/zen](https://github.com/cloudml/zen)

tdas commented on core/src/main/scala/org/apache/spark/rdd/LocalCheckpointRDD.scala:
```diff
@@ -0,0 +1,67 @@
+private[spark] class LocalCheckpointRDD[T: ClassTag](
+    @transient sc: SparkContext,
+    rddId: Int,
+    partitionIndices: Array[Int])
```
Why do you need to take partitionIndices? Arent they always going to be  0 to numPartitions-1?

tdas commented on core/src/main/scala/org/apache/spark/rdd/LocalCheckpointRDD.scala:
```diff
@@ -0,0 +1,67 @@
+  override def compute(partition: Partition, context: TaskContext): Iterator[T] = {
+    throw new SparkException(
+      s"Checkpoint block ${RDDBlockId(rddId, partition.index)} not found! Either the executor " +
+      s"that originally checkpointed this block is no longer alive, or the original RDD is " +
```
The user does not know about any "block", so they will get confused. Better to say something like "...the original data of locally checkpointed RDD is lost...you have to recompute the original RDD manually ..."

tdas commented on core/src/main/scala/org/apache/spark/rdd/LocalRDDCheckpointData.scala:
```diff
@@ -0,0 +1,83 @@
+ * An implementation of checkpointing implemented on top of Spark's caching layer.
+ *
+ * Local checkpointing trades off fault tolerance for performance by skipping the expensive
+ * step of replicating the checkpointed data in a reliable storage. Instead, checkpoint data
```
nit: ... saving the RDD data to a reliable/fault-tolerant storage.
The term "checkpoint data" is confusing. The way I reason it in my head is 
1. "RDD data" is saved to perform checkpointing
2. Once the data is saved or checkpointed, it becomes "checkpointed data". 

But this is really potayto / potawto.

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDD.scala:
```diff
@@ -1459,20 +1479,82 @@ abstract class RDD[T: ClassTag](
+   * with their cached blocks. If you must use both features, you are advised to set
+   * `spark.dynamicAllocation.cachedExecutorIdleTimeout` to a high value.
+   *
+   * The checkpoint directory set through `SparkContext#setCheckpointDir` is not read.
```
is not used.

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDD.scala:
```diff
@@ -1459,20 +1479,82 @@ abstract class RDD[T: ClassTag](
+   * useful for RDDs with long lineages that need to be truncated periodically (e.g. GraphX).
+   *
+   * Local checkpointing sacrifices fault-tolerance for performance. In particular,
+   * checkpointed data is written to ephemeral local storage (memory or disk) instead of
```
instead of ~to~ a reliable, ~distributed~ fault-tolerant storage like HDFS.

tdas replied:
...written to local storage of the executors...
(no point mentioning memory or disk, more confusing)

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDD.scala:
```diff
@@ -1459,20 +1479,82 @@ abstract class RDD[T: ClassTag](
       // children RDD partitions point to the correct parent partitions. In the future
       // we should revisit this consideration.
       RDDCheckpointData.synchronized {
```
Make whole method synchronized.

tdas commented on core/src/main/scala/org/apache/spark/rdd/ReliableCheckpointRDD.scala:
```diff
@@ -0,0 +1,194 @@
+        logInfo(s"Deleting tempOutputPath $tempOutputPath")
+        fs.delete(tempOutputPath, false)
+        throw new IOException("Checkpoint failed: failed to save output of task: " +
+          s"${ctx.attemptNumber()} and final output path does not exist")
```
Can you add the output path in the string.

tdas commented on core/src/main/scala/org/apache/spark/rdd/ReliableCheckpointRDD.scala:
```diff
@@ -0,0 +1,194 @@
+  // Test whether CheckpointRDD generate expected number of partitions despite
+  // each split file having multiple blocks. This needs to be run on a
+  // cluster (mesos or standalone) using HDFS.
+  def main(args: Array[String]) {
```
This can be removed. No one uses this. This is useless.

tdas commented on core/src/main/scala/org/apache/spark/rdd/ReliableCheckpointRDD.scala:
```diff
@@ -0,0 +1,194 @@
+  @transient private val hadoopConf = sc.hadoopConfiguration
+  @transient private val fs = new Path(checkpointPath).getFileSystem(hadoopConf)
+  private val broadcastedConf = sc.broadcast(new SerializableConfiguration(hadoopConf))
+
```
If you want, you can put a `require(fs.exists(checkpointPath))` in the constructor.

Commit:
Merge branch 'master' of github.com:apache/spark into local-checkpoint

Commit:
Address comments

AmplabJenkins commented:
Merged build triggered.

andrewor14 commented:
@witgo super cool! I'd be interested to hear back from your benchmark results.
@tdas I believe I addressed all of your comments. Please have another look.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38622 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38622/consoleFull) for   PR 7279 at commit [`c2e111b`](https://github.com/apache/spark/commit/c2e111b3ee7be3efa7dc320a09f9fa80507dd772).

SparkQA commented:
[Test build #38622 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38622/console) for   PR 7279 at commit [`c2e111b`](https://github.com/apache/spark/commit/c2e111b3ee7be3efa7dc320a09f9fa80507dd772).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(`
  * `class RFormula(override val uid: String) extends Estimator[RFormulaModel] with RFormulaBase `

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Fix compile

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38652 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38652/consoleFull) for   PR 7279 at commit [`3b7dc6a`](https://github.com/apache/spark/commit/3b7dc6ac1c175e1000fe135faa1525e06cddca83).

SparkQA commented:
[Test build #38652 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38652/console) for   PR 7279 at commit [`3b7dc6a`](https://github.com/apache/spark/commit/3b7dc6ac1c175e1000fe135faa1525e06cddca83).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38722 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38722/consoleFull) for   PR 7279 at commit [`ab003a3`](https://github.com/apache/spark/commit/ab003a38ca764d059af390796e3ff19287c7d350).

SparkQA commented:
[Test build #38722 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38722/console) for   PR 7279 at commit [`ab003a3`](https://github.com/apache/spark/commit/ab003a38ca764d059af390796e3ff19287c7d350).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
Ping @tdas can you do a final sign-off?

tdas commented on core/src/main/scala/org/apache/spark/rdd/LocalRDDCheckpointData.scala:
```diff
@@ -0,0 +1,83 @@
+
+    // Not all RDD actions compute all partitions of the RDD (e.g. take)
+    // We must compute and cache any missing partitions for correctness reasons
+    rdd.partitions.foreach { p =>
```
@andrewor14 I am a little confused about this. You are calling cachemanager in the driver, for each partition that is not present in the block manager. How does that work in the distributed manner? Dont you have to run a job (through the DAGScheduler and all) to actually run tasks?

tdas replied:
Also, is this code tested in a unit test with a action that computes partially?

andrewor14 replied:
I think you're right. We can probably replace this whole block with just an `rdd.count()` actually, since we assume that it's already marked for caching.

tdas replied:
This isnt great either. Even when its cached, it may be cached on disk, or serialized in memory. In which case running a count may be costly and time consuming, and pretty much defeats the purpose of making this a cheap checkpointing. Also, in the majority of the cases, this will be fully cached, in which case running this job is superfluous. The right thing (which aint too hard is to find out which partitions are missing and only run those partitions. 
```
val missingPartitionIds = rdd.partition.filter { p => 
   !blockManagerMaster.contains(RDDBlockId(rdd.id, p.index)) 
}.map { _.index }

rdd.sparkContext.runJob(
  rdd, 
  (tc: TaskContext, iterator: Iterator[T]) => Utils.getIteratorSize(iterator)  // same as count()
  missintPartitionIds
)
```

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDD.scala:
```diff
@@ -1448,33 +1468,92 @@ abstract class RDD[T: ClassTag](
-      }
+      checkpointData = Some(new ReliableRDDCheckpointData(this))
+    }
+  }
```
What happens when checkpointData is not empty? If it is not `ReliableRDDCheckpointData` already, the user is probably doing something wrong as he/she has called checkpoint() after calling localCheckpoint(). I think instead of being a complete no-op, we should issue a warning. Otherwise the user is going to think "I called checkpoint(), why are thinking failing"... super hard to debug.

andrewor14 replied:
ok, will add a warning

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDD.scala:
```diff
@@ -1448,33 +1468,92 @@ abstract class RDD[T: ClassTag](
+   */
+  def localCheckpoint(): this.type = RDDCheckpointData.synchronized {
+    if (conf.getBoolean("spark.dynamicAllocation.enabled", false) &&
+        conf.contains("spark.dynamicAllocation.cachedExecutorIdleTimeout")) {
```
Shoudnt this be "!conf.contains". I am guessing that you want to throw the warning when the user has enabled dynamic allocation but has not specified a timeout? Isnt it?

But even if the idle timeout is set, you dont know whether this timeout is sufficiently large for his/her purpose. So probably you should issue the warning even if the timeout is set.

andrewor14 replied:
no, the default is infinity, so it's OK if the user didn't set this.

tdas replied:
Ahh ok

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDD.scala:
```diff
@@ -1448,33 +1468,92 @@ abstract class RDD[T: ClassTag](
+      logWarning("Local checkpointing is NOT safe to use with dynamic allocation, " +
+        "which removes executors along with their cached blocks. If you must use both " +
+        "features, you are advised to set `spark.dynamicAllocation.cachedExecutorIdleTimeout` " +
+        "to a high value.")
```
nit: What is a high value? Shouldnt you give more hints.... "(e.g. if you plan to use the generated data within 1 hour, set the timeout to at least 1 hour)"

tdas commented on core/src/test/scala/org/apache/spark/CheckpointSuite.scala:
```diff
@@ -37,40 +41,42 @@ class CheckpointSuite extends SparkFunSuite with LocalSparkContext with Logging
   }
 
-  test("basic checkpointing") {
+  runTest("basic checkpointing") { normal: Boolean =>
```
What does `normal` mean!? Please rename it to something like `isLocalCheckpoint`

tdas commented on core/src/test/scala/org/apache/spark/ContextCleanerSuite.scala:
```diff
@@ -221,23 +221,26 @@ class ContextCleanerSuite extends ContextCleanerSuiteBase {
     runGC()
     postGCTester.assertCleanup()
-    assert(fs.exists(RDDCheckpointData.rddCheckpointDataPath(sc, rddId).get))
+    assert(!fs.exists(ReliableRDDCheckpointData.checkpointPath(sc, rddId).get))
```
Was this a incorrect test before?

tdas replied:
Aah, you reverse the test, first test with checkpoint cleaning enabled (as it was added to global conf), then test without checkpoint cleaning enabled. But why? Why was the global conf changed?

tdas commented on core/src/test/scala/org/apache/spark/ContextCleanerSuite.scala:
```diff
@@ -52,6 +51,7 @@ abstract class ContextCleanerSuiteBase(val shuffleManager: Class[_] = classOf[Ha
     .setAppName("ContextCleanerSuite")
     .set("spark.cleaner.referenceTracking.blocking", "true")
     .set("spark.cleaner.referenceTracking.blocking.shuffle", "true")
+    .set("spark.cleaner.referenceTracking.cleanCheckpoints", "true")
```
Why was this global conf for the test changed?

tdas replied:
I dont like the fact that the this PR touches a configuration that may affect all the tests in an unrelated testsuite. May be it is fine that there is not regression, but then one has to go through all the tests in this TestSuite to be sure.

andrewor14 replied:
This is a context cleaner suite, and its purpose is exactly to test cleaning up of various things including checkpoints. It's natural to turn this one on here since we use it in multiple tests in the suite. It's a minor point, but I really think we should just enable this here, but if you feel strongly about this I'll revert the changes.

tdas commented on core/src/test/scala/org/apache/spark/rdd/LocalCheckpointSuite.scala:
```diff
@@ -0,0 +1,330 @@
+    assert(filteredRdd.dependencies.head.rdd === mappedRdd)
+    assert(mappedRdd.dependencies.size === 1)
+    assert(mappedRdd.dependencies.head.rdd === parallelRdd)
+    assert(parallelRdd.dependencies.size === 0)
```
super nit: All these should ideally be `require` as these are making sure the assumptions are correct before the actual test.

andrewor14 replied:
I prefer to leave these as assert. We don't do require elsewhere in tests.

tdas replied:
That's fine by me. Though you have used `require` at other places even this this file.

tdas commented on core/src/test/scala/org/apache/spark/rdd/LocalCheckpointSuite.scala:
```diff
@@ -0,0 +1,330 @@
+    bmm.removeBlock(blockId)
+    try {
+      rdd.collect()
+      fail("Collect should have failed if local checkpoint block is removed...")
```
Shouldnt this be an `intercept[SparkException]`

andrewor14 replied:
no, because we want the exception object

tdas replied:
```
val exception = intercept[SparkException] {
 ....
}
```
:)

tdas commented:
I did a pass. And there is one critical question that is not clear to me -- how does the LocalCheckpointRDDData.doCheckpoint() work in a distributed manner.

Commit:
Merge branch 'master' of github.com:apache/spark into local-checkpoint

Commit:
Address comments

andrewor14 commented:
OK, I fixed the concern with the local `doCheckpoint()`. That was a good catch.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39406 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39406/consoleFull) for   PR 7279 at commit [`3be5aea`](https://github.com/apache/spark/commit/3be5aea1d93937251d0b42d7200a2421f08b9f4a).

SparkQA commented:
[Test build #39406 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39406/console) for   PR 7279 at commit [`3be5aea`](https://github.com/apache/spark/commit/3be5aea1d93937251d0b42d7200a2421f08b9f4a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `      class SpecificSafeProjection extends $`
  * `case class FromUTCTimestamp(left: Expression, right: Expression)`
  * `case class ToUTCTimestamp(left: Expression, right: Expression)`
  * `case class DateDiff(endDate: Expression, startDate: Expression)`
  * `case class InitCap(child: Expression) extends UnaryExpression with ImplicitCastInputTypes `

AmplabJenkins commented:
Merged build finished. Test FAILed.

tdas commented on core/src/test/scala/org/apache/spark/rdd/LocalCheckpointSuite.scala:
```diff
@@ -0,0 +1,330 @@
+    assert(rdd.first() === first)
+
+    // Test the same thing by calling actions on a descendant instead
+    val rdd1 = rdd.repartition(10)
```
Nit: Why would this ever fail if the `rdd.collect()` gives same result before and after checkpointing has occurred? If above you simply test whether rdd.collect() is sufficient, then these further tests seems superfluous.

andrewor14 replied:
This is testing the following case:
```
rdd.localCheckpoint().map(...).filter(...).reduceByKey(...).first()
```
where the action doesn't happen immediately after the local checkpoint. This is a real case that needs to be tested because we need to look at the last RDD's ancestors to see whether *they* are checkpointed even if the last RDD is not.

andrewor14 commented:
retest this please

SparkQA commented:
[Test build #1281 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1281/consoleFull) for   PR 7279 at commit [`3be5aea`](https://github.com/apache/spark/commit/3be5aea1d93937251d0b42d7200a2421f08b9f4a).

SparkQA commented:
[Test build #1282 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1282/consoleFull) for   PR 7279 at commit [`3be5aea`](https://github.com/apache/spark/commit/3be5aea1d93937251d0b42d7200a2421f08b9f4a).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39442 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39442/consoleFull) for   PR 7279 at commit [`3be5aea`](https://github.com/apache/spark/commit/3be5aea1d93937251d0b42d7200a2421f08b9f4a).

tdas commented on core/src/main/scala/org/apache/spark/rdd/LocalRDDCheckpointData.scala:
```diff
@@ -0,0 +1,77 @@
+    // Not all RDD actions compute all partitions of the RDD (e.g. take)
+    // We must compute and cache any missing partitions for correctness reasons
+    // TODO: avoid running another job here (SPARK-8582)
+    rdd.count()
```
I mentioned in the earlier comment thread, you may have missed it. Recommenting it here.

This `rdd.count` isnt great either. Even when its cached, it may be cached on disk, or serialized in memory. In which case running a count may be costly and time consuming, and pretty much defeats the purpose of making this a cheap checkpointing. Also, in the majority of the cases, this will be fully cached, in which case running this job is superfluous. The right thing to do (which aint too hard) is to find out which partitions are missing and only run those partitions.
```
val missingPartitionIds = rdd.partition.filter { p => 
   !blockManagerMaster.contains(RDDBlockId(rdd.id, p.index)) 
}.map { _.index }

rdd.sparkContext.runJob(
  rdd, 
  (tc: TaskContext, iterator: Iterator[T]) => Utils.getIteratorSize(iterator)  // same as count()
  missintPartitionIds
)
```

andrewor14 replied:
alright

SparkQA commented:
[Test build #1282 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1282/console) for   PR 7279 at commit [`3be5aea`](https://github.com/apache/spark/commit/3be5aea1d93937251d0b42d7200a2421f08b9f4a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1281 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1281/console) for   PR 7279 at commit [`3be5aea`](https://github.com/apache/spark/commit/3be5aea1d93937251d0b42d7200a2421f08b9f4a).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #39442 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39442/console) for   PR 7279 at commit [`3be5aea`](https://github.com/apache/spark/commit/3be5aea1d93937251d0b42d7200a2421f08b9f4a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

Commit:
Merge branch 'master' of github.com:apache/spark into local-checkpoint

Commit:
Avoid computing all partitions in local checkpoint

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39466 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39466/consoleFull) for   PR 7279 at commit [`34bc059`](https://github.com/apache/spark/commit/34bc05901b49ef46cd537242ccae4417cb015127).

SparkQA commented:
[Test build #1284 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1284/consoleFull) for   PR 7279 at commit [`34bc059`](https://github.com/apache/spark/commit/34bc05901b49ef46cd537242ccae4417cb015127).

SparkQA commented:
[Test build #1285 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1285/consoleFull) for   PR 7279 at commit [`34bc059`](https://github.com/apache/spark/commit/34bc05901b49ef46cd537242ccae4417cb015127).

SparkQA commented:
[Test build #1286 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1286/consoleFull) for   PR 7279 at commit [`34bc059`](https://github.com/apache/spark/commit/34bc05901b49ef46cd537242ccae4417cb015127).

SparkQA commented:
[Test build #1284 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1284/console) for   PR 7279 at commit [`34bc059`](https://github.com/apache/spark/commit/34bc05901b49ef46cd537242ccae4417cb015127).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class JavaKMeansExample `
  * `  class FreqSequence[Item](val sequence: Array[Array[Item]], val freq: Long) extends Serializable `
  * `class PrefixSpanModel[Item](val freqSequences: RDD[PrefixSpan.FreqSequence[Item]])`
  * `      class SpecificSafeProjection extends $`
  * `case class FromUTCTimestamp(left: Expression, right: Expression)`
  * `case class ToUTCTimestamp(left: Expression, right: Expression)`
  * `case class DateDiff(endDate: Expression, startDate: Expression)`
  * `case class InitCap(child: Expression) extends UnaryExpression with ImplicitCastInputTypes `
  * `public final class UnsafeKVExternalSorter `

SparkQA commented:
[Test build #1285 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1285/console) for   PR 7279 at commit [`34bc059`](https://github.com/apache/spark/commit/34bc05901b49ef46cd537242ccae4417cb015127).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class FreqSequence[Item](val sequence: Array[Array[Item]], val freq: Long) extends Serializable `
  * `class PrefixSpanModel[Item](val freqSequences: RDD[PrefixSpan.FreqSequence[Item]])`
  * `public final class UnsafeKVExternalSorter `

SparkQA commented:
[Test build #39466 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39466/console) for   PR 7279 at commit [`34bc059`](https://github.com/apache/spark/commit/34bc05901b49ef46cd537242ccae4417cb015127).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #1286 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1286/console) for   PR 7279 at commit [`34bc059`](https://github.com/apache/spark/commit/34bc05901b49ef46cd537242ccae4417cb015127).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

Commit:
Oops, fix tests

This proves that the test is valid!

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #1300 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1300/consoleFull) for   PR 7279 at commit [`34bc059`](https://github.com/apache/spark/commit/34bc05901b49ef46cd537242ccae4417cb015127).

SparkQA commented:
[Test build #39495 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39495/consoleFull) for   PR 7279 at commit [`729600f`](https://github.com/apache/spark/commit/729600f83072422824689fb22e5148ccccd211cf).

SparkQA commented:
[Test build #1301 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1301/consoleFull) for   PR 7279 at commit [`729600f`](https://github.com/apache/spark/commit/729600f83072422824689fb22e5148ccccd211cf).

SparkQA commented:
[Test build #1302 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1302/consoleFull) for   PR 7279 at commit [`729600f`](https://github.com/apache/spark/commit/729600f83072422824689fb22e5148ccccd211cf).

SparkQA commented:
[Test build #39495 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39495/console) for   PR 7279 at commit [`729600f`](https://github.com/apache/spark/commit/729600f83072422824689fb22e5148ccccd211cf).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RequestExecutors(appId: String, requestedTotal: Int)`
  * `  case class KillExecutors(appId: String, executorIds: Seq[String])`
  * `public class JavaKMeansExample `
  * `  class FreqSequence[Item](val sequence: Array[Array[Item]], val freq: Long) extends Serializable `
  * `class PrefixSpanModel[Item](val freqSequences: RDD[PrefixSpan.FreqSequence[Item]])`
  * `      class SpecificSafeProjection extends $`
  * `case class FromUTCTimestamp(left: Expression, right: Expression)`
  * `case class ToUTCTimestamp(left: Expression, right: Expression)`
  * `case class DateDiff(endDate: Expression, startDate: Expression)`
  * `case class InitCap(child: Expression) extends UnaryExpression with ImplicitCastInputTypes `
  * `public final class UnsafeKVExternalSorter `

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #1301 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1301/console) for   PR 7279 at commit [`729600f`](https://github.com/apache/spark/commit/729600f83072422824689fb22e5148ccccd211cf).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1300 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1300/console) for   PR 7279 at commit [`34bc059`](https://github.com/apache/spark/commit/34bc05901b49ef46cd537242ccae4417cb015127).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #1302 timed out](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1302/console)**     for PR 7279 at commit [`729600f`](https://github.com/apache/spark/commit/729600f83072422824689fb22e5148ccccd211cf)     after a configured wait of `175m`.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39536 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39536/consoleFull) for   PR 7279 at commit [`729600f`](https://github.com/apache/spark/commit/729600f83072422824689fb22e5148ccccd211cf).

SparkQA commented:
[Test build #39536 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39536/console) for   PR 7279 at commit [`729600f`](https://github.com/apache/spark/commit/729600f83072422824689fb22e5148ccccd211cf).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #1304 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1304/consoleFull) for   PR 7279 at commit [`729600f`](https://github.com/apache/spark/commit/729600f83072422824689fb22e5148ccccd211cf).

SparkQA commented:
[Test build #1304 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1304/console) for   PR 7279 at commit [`729600f`](https://github.com/apache/spark/commit/729600f83072422824689fb22e5148ccccd211cf).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

andrewor14 commented:
@tdas good to go?

tdas commented:
Yep, LGTM. Merging this to master. Great patch!! Thanks!
